File: src/main/java/com/epam/eng/pattern/observer/WeatherStation.java
package com.epam.eng.pattern.observer;

public class WeatherStation {
    private int temperature;
    private int humidity;
    private int pressure;

    public void setTemperature(int temperature) {
        this.temperature = temperature;
    }

    public void setHumidity(int humidity) {
        this.humidity = humidity;
    }

    public void setPressure(int pressure) {
        this.pressure = pressure;
    }

    public void measureWeather() {

        int temperature = getTemperatureFromSensor();
        int humidity = getHumidityFromSensor();
        int pressure = getPressureFromSensor();

        setTemperature(temperature);
        setHumidity(humidity);
        setPressure(pressure);
    }

    private int getPressureFromSensor() {
        return 1;
    }

    private int getHumidityFromSensor() {
        return 1;
    }

    private int getTemperatureFromSensor() {
        return 1;
    }

}

File: ts/observer/WeatherStation.ts
interface Observer {
  update(temperature: number, humidity: number, pressure: number): void;
}

interface Subject {
  registerObserver(o: Observer): void;
  removeObserver(o: Observer): void;
  notifyObservers(): void;
}

class WeatherStation implements Subject {
  private temperature: number;
  private humidity: number;
  private pressure: number;
  private observers: Observer[] = [];

  public registerObserver(o: Observer): void {
    this.observers.push(o);
  }

  public removeObserver(o: Observer): void {
    const index = this.observers.indexOf(o);
    if (index !== -1) {
      this.observers.splice(index, 1);
    }
  }

  public notifyObservers(): void {
    for (let observer of this.observers) {
      observer.update(this.temperature, this.humidity, this.pressure);
    }
  }

  public setMeasurements(
    temperature: number,
    humidity: number,
    pressure: number
  ): void {
    this.temperature = temperature;
    this.humidity = humidity;
    this.pressure = pressure;
    this.notifyObservers();
  }

  public measureWeather(): void {
    const temperature: number = this.getTemperatureFromSensor();
    const humidity: number = this.getHumidityFromSensor();
    const pressure: number = this.getPressureFromSensor();

    this.setMeasurements(temperature, humidity, pressure);
  }

  private getPressureFromSensor(): number {
    return 1;
  }

  private getHumidityFromSensor(): number {
    return 1;
  }

  private getTemperatureFromSensor(): number {
    return 1;
  }
}

class WeatherDisplay implements Observer {
  private temperature: number;
  private humidity: number;
  private pressure: number;

  public update(temperature: number, humidity: number, pressure: number): void {
    this.temperature = temperature;
    this.humidity = humidity;
    this.pressure = pressure;
    this.display();
  }

  public display(): void {
    console.log(
      `Temperature: ${this.temperature}, Humidity: ${this.humidity}, Pressure: ${this.pressure}`
    );
  }
}
